packages/commerce-sdk-react/src/auth/index.ts

```diff
@@ -0,0 +1,297 @@
+import {helpers, ShopperLogin, ShopperLoginTypes} from 'commerce-sdk-isomorphic'
+import jwtDecode from 'jwt-decode'
+import {ApiClientConfigParams} from '../hooks/types'
+import {BaseStorage, LocalStorage, CookieStorage} from './storage'
+
+type Helpers = typeof helpers
+interface AuthConfig {
+    shortCode: string
+    organizationId: string
+    clientId: string
+    siteId: string
+    redirectURI: string
+    proxy: string
+}
+
+interface JWTHeaders {
+    exp: number
+    iat: number
+}
+
+// this type is slightly different from ShopperLoginTypes.TokenResponse, reasons:
+// 1. TokenResponse is too generic (with & {[key:string]: any}), we need a more
+//    restrictive type to make sure type safe
+// 2. The refresh tokens are stored separately for guest and registered user. Instead
+//    of refresh_token, we have refresh_token_guest and refresh_token_registered
+type AuthDataKeys =
+    | 'access_token'
+    | 'customer_id'
+    | 'enc_user_id'
+    | 'expires_in'
+    | 'id_token'
+    | 'idp_access_token'
+    | 'refresh_token_guest'
+    | 'refresh_token_registered'
+    | 'token_type'
+    | 'usid'
+type AuthDataMap = Record<
+    AuthDataKeys,
+    {
+        storage: BaseStorage
+        key: string
+        callback?: () => void
+    }
+>
+
+/**
+ *
+ * @Internal
+ */
+class Auth {
+    private client: ShopperLogin<ApiClientConfigParams>
+    private redirectURI: string
+    private _onClient = typeof window !== 'undefined'
+    private pending: Promise<ShopperLoginTypes.TokenResponse> | undefined
+    private localStorage: LocalStorage
+    private cookieStorage: CookieStorage
+    private REFRESH_TOKEN_EXPIRATION_DAYS = 90
+
+    static COOKIE = 'COOKIE'
+    static LOCAL_STORAGE = 'LOCAL_STORAGE'
+
+    constructor(config: AuthConfig) {
+        this.client = new ShopperLogin({
+            proxy: config.proxy,
+            parameters: {
+                clientId: config.clientId,
+                organizationId: config.organizationId,
+                shortCode: config.shortCode,
+                siteId: config.siteId,
+            },
+        })
+
+        this.redirectURI = config.redirectURI
+        this.localStorage = this._onClient ? new LocalStorage() : new Map()
+        this.cookieStorage = this._onClient ? new CookieStorage() : new Map()
```

Comment:
There is going to be some work in the future to share guest auth on the server. I don't know how this is going to work, but I can image that we'd be storing the auth in some kinda of globally accessible location. 

Do you think we should make the storage something that is configurable?

Reply:
I think we eventually should, but I'm not sure about it for the initial release thou. I'd like to see actual use cases before designing the configuration model...